pkg/volume/flocker/client.go

```diff
@@ -0,0 +1,391 @@
+	defaultVolumeSize = json.Number("107374182400")
+
+	// Flocker connections are authenticated with TLS
+	// TODO: It can perhaps be stored somewhere else.
+	defaultCACertFile     = "/etc/flocker/cluster.crt"
+	defaultClientKeyFile  = "/etc/flocker/apiuser.key"
+	defaultClientCertFile = "/etc/flocker/apiuser.crt"
+)
+
+var (
+	// A volume can take a long time to be available, if we don't want
+	// Kubernetes to wait forever we need to stop trying after some time, that
+	// time is defined here
+	timeoutWaitingForVolume = 2 * time.Minute
+	tickerWaitingForVolume  = 5 * time.Second
+
+	errStateNotFound         = errors.New("State not found by Dataset ID")
+	errConfigurationNotFound = errors.New("Configuration not found by Name")
+)
+
+func newTLSClient(caCertPath, keyPath, certPath string) (*http.Client, error) {
+	// Client certificate
+	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// CA certificate
+	caCert, err := ioutil.ReadFile(caCertPath)
+	if err != nil {
+		return nil, err
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      caCertPool,
+	}
+	tlsConfig.BuildNameToCertificate()
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+	return &http.Client{Transport: transport}, nil
+}
+
+type httpDoer interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// TODO: use this
+type flockerClientable interface {
+	httpDoer
+	CreateVolume(string) (string, error)
+}
+
+type flockerClient struct {
+	*http.Client
+
+	pod *api.Pod
+
+	schema  string
+	host    string
+	port    int
+	version string
+
+	maximumSize json.Number
+}
+
+var (
+	errFlockerControlServiceHost = errors.New("The environment variable FLOCKER_CONTROL_SERVICE_HOST can't be empty")
+	errFlockerControlServicePort = errors.New("The environment variable FLOCKER_CONTROL_SERVICE_PORT must be a number")
+)
+
+func newFlockerClientWithoutSize(pod *api.Pod) (*flockerClient, error) {
+	return newFlockerClient(pod, true)
+}
+
+func newFlockerClientWithSize(pod *api.Pod, maximumSize int64) (*flockerClient, error) {
+	c, err := newFlockerClient(pod, true)
+	if err != nil {
+		return nil, err
+	}
+	c.maximumSize = json.Number(string(maximumSize))
+	return c, err
+}
+
+/*
+ * newFlockerClient creates a wrapper over http.Client to communicate with the
+ * flocker control service. The location of this service is defined by the
+ * following environment variables:
+ *
+ * - FLOCKER_CONTROL_SERVICE_HOST
+ * - FLOCKER_CONTROL_SERVICE_PORT
+ *
+ * If you are using the TLS client you can also specify where the cert files
+ * reside:
+ *
+ * - FLOCKER_CONTROL_SERVICE_CA_FILE
+ * - FLOCKER_CONTROL_SERVICE_CLIENT_KEY_FILE
+ * - FLOCKER_CONTROL_SERVICE_CLIENT_CERT_FILE
+ */
+func newFlockerClient(pod *api.Pod, useTLS bool) (*flockerClient, error) {
+	host := os.Getenv("FLOCKER_CONTROL_SERVICE_HOST")
```

Comment:
If this is configured once for the plugin and its value is correct across the cluster, then using VolumeConfig would allow CLI configuration of all these variables.